operator/cmd/mesh/manifest-validate.go

```diff
@@ -0,0 +1,345 @@
+// Copyright 2020 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mesh
+
+import (
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/gogo/protobuf/types"
+
+	"istio.io/istio/operator/pkg/util"
+	"istio.io/istio/operator/pkg/validate"
+	"istio.io/istio/pkg/config/validation"
+)
+
+const (
+	traceSamplingMin float64 = 0.0
+	traceSamplingMax float64 = 100.0
+)
+
+var (
+	// Boolean values for --set flags
+	boolValues = []bool{true, false}
+
+	// Ref: https://kubernetes.io/docs/concepts/configuration/overview/#container-images
+	imagePullPolicy = []string{"Always", "IfNotPresent", "Never"}
+
+	// Keep this list updated as per following
+	// https://preliminary.istio.io/docs/setup/additional-setup/config-profiles/
+	profile = []string{"default", "demo", "empty", "minimal", "preview", "remote", "separate"}
+
+	inboundClusterStatName = []string{"SERVICE", "SERVICE_FQDN", "SERVICE_PORT", "SERVICE_PORT_NAME"}
+
+	outboundClusterStatName = append(inboundClusterStatName, "SUBSET_NAME")
+
+	// https://preliminary.istio.io/docs/reference/config/istio.mesh.v1alpha1.html#MeshConfig-OutboundTrafficPolicy-Mode
+	outboundTrafficPolicyMode = []string{"REGISTRY_ONLY", "ALLOW_ANY"}
+
+	// https://preliminary.istio.io/docs/reference/config/istio.mesh.v1alpha1.html#MeshConfig-H2UpgradePolicy
+	h2UpgradePolicy = []string{"DO_NOT_UPGRADE", "UPGRADE"}
+
+	// https://preliminary.istio.io/docs/reference/config/istio.operator.v1alpha1/#IstioOperatorSpec
+	setFlagValues = map[string]interface{}{
+		"profile": profile,
+
+		"installPackagePath": validate.InstallPackagePath,
+		"hub":                validate.Hub,
+		"tag":                validate.Tag,
+
+		"namespace": validate.CheckNamespaceName,
+		"revision":  validate.CheckRevision,
+
+		// MeshConfig related flags
+		// https://preliminary.istio.io/docs/reference/config/istio.mesh.v1alpha1.html#MeshConfig
+		"values.global.mixerCheckServer":                validation.ValidateProxyAddress,
+		"values.global.mixerReportServer":               validation.ValidateProxyAddress,
+		"values.global.disablePolicyChecks":             boolValues,
+		"values.global.policyCheckFailOpen":             boolValues,
+		"values.mixer.telemetry.sessionAffinityEnabled": boolValues,
+		"values.meshConfig.proxyListenPort":             validatePort,
+		"values.meshConfig.proxyHttpPort":               validatePort,
```

Comment:
We can (and should) get all these types directly from the go structs (you can use tpath functions to reflect through the structs to get the types, GetPathContext will return an interface{} of the struct path that you can get the type from).
Maintaining this type of table is not feasible in the long run.

Reply:
> This won't work properly because the allowed values depend on the scanned file list in operator/data/profiles.

Thanks, As you mentioned, reading profiles from `operator/data/profiles`, profile list will be dynamically updated.

Reply:
> We can (and should) get all these types directly from the go structs (you can use tpath functions to reflect through the structs to get the types, GetPathContext will return an interface{} of the struct path that you can get the type from).
> Maintaining this type of table is not feasible in the long run.

I agree, But I'm not clear how to use `tpath` here. Can you please provide me some pointers to better understand this.

Reply:
Sure, let me try to explain a bit more. What you are wanting to do here is to figure out the type of some path in IstioOperator e.g. values.meshConfig.proxyListenPort. 
Rather than doing this statically as you have here, you can use tpath.GetPathContext to check the type. What I'd do is actually recurse through the input YAML tree and for each path, use tpath.GetPathContext to learn the type and do the validation against that. You'd need to have some unmarshaled IstioOperator object to traverse through and that could be the default profile, which sets most of the fields.

Reply:
@ostromart 
I have a limited understanding of different operator objects and how they work together. But as per my findings, the default profile is good for very basic IstioOperator objects [but lacks most of the flags](https://github.com/istio/istio/issues/20730#issuecomment-608523833). Consider the example of [MeshConfig](https://istio.io/docs/reference/config/istio.mesh.v1alpha1/#MeshConfig) flags. The default profile[ sets only one flag for mesh config](https://github.com/istio/istio/blob/master/manifests/profiles/default.yaml#L10).

If we consider the default profile to traverse the unmarshaled IstioOperator object, we will miss validation for lot of flags. We need some way where the type for particular flag is already identified so we can call a specific function to validate the values provided in `manifest apply`.

Consider the following example of fields with durations types. None of them are set in [the default profile](https://github.com/istio/istio/blob/master/manifests/profiles/default.yaml) 

- values.global.[connectTimeout](https://github.com/istio/istio/blob/master/pkg/config/validation/validation.go#L1121)
- values.global.proxy.[protocolDetectionTimeout](https://github.com/istio/istio/blob/master/pkg/config/validation/validation.go#L1132)
- values.global.proxy.[dnsRefreshRate](https://github.com/istio/istio/pull/22799/files#diff-fe3f41aa20523a21cf2e4f02c1b38519R2904)
- values.mixer.telemetry.[reportBatchMaxTime](https://github.com/istio/istio/pull/22799/files#diff-fe3f41aa20523a21cf2e4f02c1b38519R2916)

But all of them require specific validation rules even though they all are of duration type. So somehow we need 1:1 mapping of these functions which will be called after parsing `manifest apply`, identifying it's a type and redirecting it to specific validation rule.

[This table](https://github.com/istio/istio/pull/22799/files#diff-2405f93b84e56f54e6d8c6052b918e52R82) is not only to identify type but also to map to specific validation functions. It also [maintains the context](https://github.com/istio/istio/issues/22286#issuecomment-601168945) of the error [which was previously difficult to track](https://github.com/istio/istio/issues/22286). The good thing is static validation utilizes more than 60% of functions from semantic validation rules. (see links above)

Please let me know if any of these sounds resonable.
CC: @esnible @howardjohn

Reply:
@ostromart @esnible @howardjohn can you PTAL ?